AnkiDroid/src/main/java/com/ichi2/anki/account/LoginViewModel.kt

```diff
@@ -0,0 +1,125 @@
+    private val _loginButtonEnabled = MutableStateFlow(false)
+    val loginButtonEnabled: StateFlow<Boolean> = _loginButtonEnabled.asStateFlow()
+
+    private val _userNameError = MutableStateFlow<LoginError?>(null)
+    val userNameError: StateFlow<LoginError?> = _userNameError.asStateFlow()
+
+    private val _passwordError = MutableStateFlow<LoginError?>(null)
+    val passwordError: StateFlow<LoginError?> = _passwordError.asStateFlow()
+
+    private val _loginState = MutableStateFlow<LoginState>(LoginState.Idle)
+    val loginState: StateFlow<LoginState> = _loginState.asStateFlow()
+
+    fun onUserNameFocusChange(
+        hasFocus: Boolean,
+        userName: String,
+    ) {
+        _userNameError.value = if (!hasFocus && userName.isEmpty()) LoginError.EMPTY_USERNAME else null
+    }
+
+    fun onPasswordFocusChange(
+        hasFocus: Boolean,
+        password: String,
+    ) {
+        _passwordError.value = if (!hasFocus && password.isEmpty()) LoginError.EMPTY_PASSWORD else null
+    }
+
+    fun onTextChanged(
+        userName: String,
+        password: String,
+    ) {
+        _loginButtonEnabled.value = userName.isNotEmpty() && password.isNotEmpty()
+    }
+
+    /**
+     * Handles the login process by attempting to authenticate the user with the provided credentials.
+     * If authentication is successful, it updates the login state; otherwise, it handles authentication failure.
+     *
+     * @param username The username entered by the user.
+     * @param password The password entered by the user.
+     * @param endpoint An endpoint for authentication.
+     * @param syncLogin A suspend function that performs the authentication and returns a {@link SyncAuth} object.
+     * @param updateLogin A function that updates the login credentials upon successful authentication.
+     */
+    fun handleLogin(
+        username: String,
+        password: String,
+        endpoint: String?,
+        syncLogin: suspend (String, String, String?) -> SyncAuth,
+        updateLogin: (String, String) -> Unit,
+    ) {
+        viewModelScope.launch {
+            try {
+                val auth = syncLogin(username, password, endpoint)
+                updateLogin(username, auth.hkey)
+                _loginState.value = LoginState.Success
+            } catch (exc: BackendSyncException.BackendSyncAuthFailedException) {
+                updateLogin("", "")
+                _loginState.value = LoginState.Error(exc)
+            }
+        }
+    }
+}
+
+/**
+ * Enum representing the possible error states for the login fields.
+ * It is used to track errors related to the username and password fields.
+ */
+enum class LoginError {
+    EMPTY_USERNAME,
+    EMPTY_PASSWORD,
+    ;
+
+    fun toHumanReadableString(context: Context): String? =
+        when (this) {
+            EMPTY_USERNAME -> context.getString(R.string.invalid_email)
+            EMPTY_PASSWORD -> context.getString(R.string.password_empty)
+        }
+}
```

Comment:
This seems less repetitive
```suggestion
enum class LoginError(@StringRes private val messageResId: Int) {
    EMPTY_USERNAME(R.string.invalid_email),
    EMPTY_PASSWORD(R.string.password_empty),
    ;

    fun toHumanReadableString(context: Context): String? =
        context.getString(this.messageResId)
}
```